README.md
# aruba-central-ansible-role

This Ansible Network role provides a set of platform dependent configuration management modules and plugins specifically designed for Aruba Central, which is a cloud-based Network Management System.

## [](https://github.com/aruba/aruba-central-ansible-role#requirements)Requirements

* Python 3.5+
* Ansible 2.9 or later  
  * Ansible 2.10+ requires `ansible.netcommon` collection to be installed
* Minimum supported Aruba Central firmware is 2.5.2

* Install all Ansible requirements, with the following command:
    ```
    ansible-galaxy install -r requirements.yml
    ```
* Install all Python requirements with the following command:
    ```
    pip install -r requirements.txt
    ```
## [](https://github.com/aruba/aruba-central-ansible-role#installation)Installation

Through Galaxy:

```
ansible-galaxy install arubanetworks.aruba_central_role
```

Through Github, use the following command. Use option  `-f`  to overwrite current role version:

```
ansible-galaxy install git+https://github.com/aruba/aruba-central-ansible-role.git
```
## [](https://github.com/aruba/aruba-central-ansible-role#notes)Notes

- The modules in this role use Central's REST API. For information on REST API and how to obtain an access for using REST API, visit the Aruba Developer Hub: [Getting Started with REST API](https://developer.arubanetworks.com/aruba-central/docs/api-getting-started)
- An API token must be created for a user on Aruba Central's API Gateway and a valid, non-expired **`access_token`** must be used. For more information on how to get started with the API Gateway you can also watch this [YouTube Video](https://www.youtube.com/watch?v=tWEsL7zSOB0).
- A valid access token can be used in an Inventory file as mentioned in the [Inventory](https://github.com/aruba/aruba-central-ansible-role#inventory) section.
- Ensure that the access token was created with "Network Operations" selected in the Application drop-down list, while adding a new token on the API Gateway. 
- Once a new token is generated, it will have an `access_token` and a `refresh_token`.
- An access token is valid for a period of 7200 seconds or two hours. After two hours, it will expire and a new token needs to be created. The token expiry time is currently not configurable.
- The **`refresh_token`** along with the **`client_id`** and **`client_secret`** are used to renew the access token. This functionality is implemented in this role by using an Inventory Plugin. You can either use an Inventory file or use an Inventory Plugin Config file for using the plugin.
- More details on how to use this plugin config file with tokens and other credentials for auto-renew of tokens is mentioned below in the [Inventory Plugin Config File](https://github.com/aruba/aruba-central-ansible-role#inventory-plugin-config-file) section.

## [](https://github.com/aruba/aruba-central-ansible-role#inventory) Inventory/Host File
There are two ways in which you can use an inventory or host file with the Aruba Central Ansible Role:
1. **Inventory** 
	    - Host file which tells Ansible the required httpapi plugin to use with some other details and the access token.
2. **Inventory Plugin Config File** or **Inventory Source**
	- An inventory source file which is used by the inventory plugin.
	- A typical inventory plugin implementation has a plugin script (usually written in Python), and an inventory source (in this case it is a YAML file).
	- Based on Ansible Documentation,  Inventory sources are the input strings that inventory plugins work with. An inventory source can be a path to a file or to a script, or it can be raw data that the plugin can interpret to dynamically generate inventory variables.
    - Only `.yml` files with [Inventory Plugin Config File variables](https://github.com/aruba/aruba-central-ansible-role#inventory-plugin-config-variables) are accepted as an inventory source by the inventory plugin for this role.
- You can use either of the above files depending on your need. 
	- If do not require auto-renew of tokens, use the simple [Inventory file](https://github.com/aruba/aruba-central-ansible-role#sample-inventory)
	- Or, if you wish to have the auto-renew of tokens, use an [Inventory Plugin Config File](https://github.com/aruba/aruba-central-ansible-role#sample-inventory-plugin-config-file) for the Inventory Plugin.


## Inventory 
### [](https://github.com/aruba/aruba-cnetral-ansible-role#inventory-variables)Inventory Variables

The variables that should be defined in your inventory for your Aruba Central account are:

- `ansible_host`: Cluster-specific Base-URL for the API Gateway on Aruba Central in FQDN format, which can be found in the API Documentation URL on API Gateway
- `ansible_connection`: Must always be set to  `httpapi`
- `ansible_network_os`: Must always be set to  `aruba_central`
- `ansible_httpapi_use_ssl`: Must always be set to  `True`
- `ansible_httpapi_central_access_token`: Aruba Central's API access token

#### [](https://github.com/aruba/aruba-central-ansible-role#sample-inventory)Sample Inventory:

##### YAML

```YAML
all:
  hosts:
    central:
      ansible_host: apigw-prod2.central.arubanetworks.com
      ansible_connection: httpapi
      ansible_network_os: aruba_central
      ansible_httpapi_use_ssl: True
      ansible_httpapi_central_access_token: CnjDaXXxvnjrvJRwxxxxXXxxXXXXxxxx
```

##### INI

```INI
arubacentral ansible_host=apigw-prod2.central.arubanetworks.com  ansible_connection=httpapi ansible_network_os=aruba_central  ansible_httpapi_use_ssl=True  ansible_httpapi_central_access_token=CnjDaXXxvnjrvJRwxxxxXXxxXXXXxxxx
```

## Inventory Plugin Config File
- This is an inventory source file which is utilized by the inventory plugin to  dynamically generate an inventory  with all the options (as shown above in the simple Inventory file) required by the HttpAPI Connection Plugin.
- User needs to create a config file which has the plugin name and other central credentials.

### Caveats
- Conventionally, an Inventory Plugin can not be shipped within a role since Ansible executes the Inventory Plugin before execution of the Playbook or Role.  More information can be found on [Ansible Docs](https://docs.ansible.com/ansible/latest/dev_guide/developing_inventory.html#inventory-plugins).
	- Therefore, until we publish the Aruba Central Ansible Collection, you need to perform the following workaround for using the custom inventory plugin for this Ansible role for Central - which will take care of the auto-renew of tokens.

**Method 1:**
- Copy the `central_inventory.py` inventory plugin file  from [GitHub](https://raw.githubusercontent.com/aruba/aruba-central-ansible-role/master/inventory_plugins/central_inventory.py) and store it within an `inventory_plugins` directory in your playbooks directory. 
- Your playbooks directory should have the following structure:
```
playbooks_dir
+-- playbook1.yml
+-- playbook2.yml
+-- inv_src.yml
+-- inventory_plugins/
|   +-- central_inventory.py
```
- Where `**inv_src.yml**` or any other `**.yml**` file with a different name can act as the Inventory Plugin Config file. [Sample Inventory Plugin Config File](https://github.com/aruba/aruba-central-ansible-role#inventory-plugin-config-variables) and the variables it uses are given below.
- Inventory Plugin Config File should not be used with Ansible Vault since the inventory plugin needs to write the renewed refresh and access tokens back to the plugin config file.
- Initially the user has to ensure that a valid Access and Refresh Token has been entered in the Inventory Plugin Config file for the first time. If both tokens are invalid, the inventory plugin will modify the file with **<Enter a Valid Access/Refresh Token>** message in the Inventory Plugin Config file and the execution will fail with "Unauthorized" failure.
- Central's **`refresh_token`** is valid for a period of **14 days**. If not used until 14 days, the token will be revoked and a new token has to be created. Refresh token validity is non-configurable at the moment.

**Method 2:**
- Once the role is installed, go to the roles directory. Use the following commands to get the path where role is installed and move inventory_plugins directory with the plugin file.
 ```
$ ansible-galaxy role list 

------------------output-----------------
# /home/admin/.ansible/roles
- arubanetworks.aruba_central_role, 0.2.1
 ``` 
- Select the path and go to the roles directory
- The role name might differ based on how you chose to install it
- Once you have the path move the inventory plugin directory to your playbooks directory
```
$ cd /home/admin/.ansible/roles

$ cd arubanetworks.aruba_central_role
or
$ cd aruba-central-ansible-role

$ mv inventory_plugins/ <path_to_playbooks_directory>
```
### [](https://github.com/aruba/aruba-central-ansible-role#inventory-plugin-config-variables)Inventory Plugin Config Variables

The variables that should be defined in your inventory plugin config file for your Aruba Central account are:

- `access_token`: Aruba Central's API Access Token.
- `api_gateway`: Cluster-specific Base-URL for the API Gateway on Aruba Central in FQDN format, which can be found in the API Documentation URL on API Gateway
- `client_id`: Aruba Central's API Client ID
- `client_secret`: Aruba Central's API Client Secret
- `host`: Must always be set to  `central`
- `plugin`: Must always be set to  `central_inventory`
- `refresh_token`: Aruba Central's API Refresh token



#### [](https://github.com/aruba/aruba-central-ansible-role#sample-inventory-plugin-config-file)Sample Inventory Plugin Config File:

##### YAML

```YAML
access_token: CnjDaXXxvnjrvJRwxxxxXXxxXXXXxxxx
api_gateway: apigw-prod2.central.arubanetworks.com
client_id: FOqWxx124ASdfS36HqKIeXXzZ
client_secret: O2RfdKgiS13GhswdrWAIEueMPOxxZxX
host: central
plugin: central_inventory
refresh_token: X12daE6BFhk8QqqzzeifHTYxxZZ12XxX

```

## [](https://github.com/aruba/aruba-central-ansible-role#example-playbook)Example Playbooks

### Including the Role

If role installed through [Galaxy](https://galaxy.ansible.com/arubanetworks/aruba_central_role)  set role to  `arubanetworks.aruba_central_role`:

    ---
    -  hosts: all
       roles:
         - role: arubanetworks.aruba_central_role
       tasks:
       - name: Get all the UI and Template Groups on Central
         central_groups:
           action: get_groups
           limit: 20
           offset: 0
If role installed through [Github](https://github.com/aruba/aruba-central-ansible-role)  set role to  `aruba-central-ansible-role`:

    ---
    -  hosts: all
       roles:
         - role: aruba-central-ansible-role
       tasks:
       - name: Get all the UI and Template Groups on Central
         central_groups:
           action: get_groups
           limit: 20
           offset: 0

  
## [](https://github.com/aruba/aruba-central-ansible-role#playbook-execution) Playbook Execution

```
ansible-playbook playbook.yml -i inventory.yml
```
- Where `inventory.yml` could either be a simple inventory file or it could be an inventory plugin config file (inventory source).
- Make sure to have the [`central_inventory.py`](https://raw.githubusercontent.com/aruba/aruba-central-ansible-role/master/inventory_plugins/central_inventory.py) in the `inventory_plugins/` directory before executing the playbook using the inventory plugin config file.

Contribution
-------

At Aruba Networks we're dedicated to ensuring the quality of our products, so if you find any
issues at all please open an issue on our [Github](https://github.com/aruba/aruba-central-ansible-role) and we'll be sure to respond promptly!

For more contribution opportunities follow our guidelines outlined in our [CONTRIBUTING.md](https://github.hpe.com/switchautomation/aruba-central-ansible-role/blob/master/CONTRIBUTING.md)

License
-------

MIT

Author Information
------------------

- Jay Pathak (@jayp193)
- Derek Wang (@derekwangHPEAruba)


RELEASE-NOTES.md
# 0.1.0

## Notable Changes
* This is the initial release for the Aruba Central Ansible Role.


# 0.2.0

## Notable Changes
* Bug fixes in the module files related to Python3
* Added requirements.yml to install ansible.netcommon for Ansible 2.10+


# 1.0.0

## Notable Changes
* Added new inventory plugin for auto-renewing access token for Aruba Central
* Added a sample inventory source file (aka inventory plugin config file)
* Modified Readme and other documentation explaining the use of this inventory plugin


library/central_groups.py
#!/usr/bin/python
'''
Central Groups Ansible Module
'''
# MIT License
#
# Copyright (c) 2020 Aruba, a Hewlett Packard Enterprise company
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.0',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = """
---
module: central_groups
version_added: 2.9.0
short_descriptions: REST API module for groups on Aruba Central
description: This module provides a mechanism to interact with groups used for
             configuration management of devices on Aruba Central.
options:
    action:
        description:
            - Action to be performed on the group(s)
            - "get_groups" gets names of existing groups
            - "get_group_mode" gets group modes of existing groups
            - "clone" creates a new group by cloning an existing group
            - "update" updates the group password of an existing UI group
            - "create" creates a new group
            - "delete" deletes an existing group
        required: true
        type: str
        choices:
            - get_groups
            - get_group_mode
            - clone
            - update
            - create
            - delete
    group_name:
        description:
            - Name of the group
            - Used with actions "clone", "update", "create", and "delete"
        required: false
        type: str
    group_list:
        description:
            - List of group names
            - At most 20 names can be listed
            - Used with action "get_group_mode"
        required: false
        type: list
    group_attributes:
        description:
            - Group attributes to define group password and a boolean
              determining whether it is a template group or UI group OR
              dictionary containing two booleans to determine the allowable
              templates (wired and/or wireless) for the template group
            - Used with actions "create" and "update"
        required: false
        type: dict
    clone_from_group:
        description:
            - Name of existing group from which the new group is cloned
            - Used with action "clone"
        required: false
        type: str
    limit:
        description:
            - Maximum number of records to be returned
            - Used optionally as a filter parameter for "get_groups"
        required: false
        type: int
        default: 20
    offset:
        description:
            - Number of items to be skipped before returning the data, which
              is useful for pagination
            - Used optionally as a filter parameter for "get_groups"
        required: false
        type: int
        default: 0

"""
EXAMPLES = """
#Usage Examples
- name: Get all the UI and template groups on Central
  central_groups:
    action: get_groups
    limit: 20
    offset: 0

- name: Get groups' configuration modes ("UI" or "template")
  central_groups:
    action: get_group_mode
    group_list:
      - group-name-1
      - group-name-2

- name: Create a new group ("template" for wired, "UI" for wireless)
  central_groups:
    action: create
    group_name: new-test-group
    group_attributes:
      group_password: admin@12345
      template_group:
        wired: True
        wireless: False

- name: Update an existing group (only available for UI groups)
  central_groups:
    action: update
    group_name: new-test-group
    group_attributes:
      group_password: Aruba@2222
      template_group: False

- name: Clone an existing group
  central_groups:
    action: clone
    group_name: new-group
    clone_from_group: new-test-group

- name: Delete a group
  central_groups:
    action: delete
    group_name: new-test-group

"""

import json  # NOQA
from ansible.module_utils.basic import AnsibleModule  # NOQA
from ansible.module_utils.central_http import CentralApi  # NOQA


def error_msg(method):
    '''
    Error handler for errors related to missing playbook parameters for groups
    module
    '''
    result = {"resp": None, "code": 400}
    if method == "create" or method == "update":
        resp = "Group name or Group attributes not present in the playbook"
    elif method == "delete":
        resp = "Group name to be deleted not present in the playbook"
    elif method == "clone":
        resp = "Group name or clone-from-group parameters not present in " \
               "the playbook"
    elif method == "get":
        resp = "List of groups (group_list) not present in the playbook"
    result['resp'] = resp
    return result


def get_groups(central_api, limit, offset):
    '''
    Gets names of existing Central Groups
    '''
    endpoint = "/configuration/v2/groups"
    query_params = {"limit": limit, "offset": offset}
    headers = central_api.get_headers(False, "get")
    path = central_api.get_url(endpoint, query_params)
    result = central_api.get(path=path, headers=headers)
    return result


def get_group_mode(central_api, group_list):
    '''
    Gets group modes of existing Central Groups
    '''
    if group_list is not None:
        groups = central_api.get_list_params(group_list)
        query_params = {"groups": groups}
        headers = central_api.get_headers(False, "get")
        path = central_api.get_url("/configuration/v2/groups/template_info",
                                   query_params)
        result = central_api.get(path=path, headers=headers)
        return result
    return error_msg("get")


def clone(central_api, group_name, clone_from_group):
    '''
    Creates a new group by cloning an existing group
    '''
    if group_name and clone_from_group is not None:
        path = "/configuration/v2/groups/clone"
        data = {"group": group_name, "clone_group": clone_from_group}
        headers = central_api.get_headers(False, "post")
        result = central_api.post(path=path, headers=headers, data=data)
        return result
    return error_msg("clone")


def create_group(central_api, group_name, group_attributes):
    '''
    Creates a new template or UI group based on group_attributes
    '''
    if group_name and group_attributes is not None:
        path = "/configuration/v2/groups"
        data = {
            "group": group_name,
            "group_attributes": {
                "group_password": group_attributes['group_password'],
                "template_info": {
                    "Wired": group_attributes['template_group']['wired'],
                    "Wireless": group_attributes['template_group']['wireless']
                }}}
        headers = central_api.get_headers(False, "post")
        result = central_api.post(path=path, headers=headers, data=data)
        return result
    return error_msg("create")


def update_group(central_api, group_name, group_attributes):
    '''
    Updates an existing UI group to change its password
    '''
    if group_name and group_attributes is not None:
        path = "/configuration/v1/groups/"+str(group_name)
        data = group_attributes
        headers = central_api.get_headers(False, "post")
        result = central_api.patch(path=path, headers=headers, data=data)
        return result
    return error_msg("update")


def delete_group(central_api, group_name):
    '''
    Deletes an existing group
    '''
    if group_name is not None:
        path = "/configuration/v1/groups/"+str(group_name)
        headers = central_api.get_headers(False, "delete")
        result = central_api.delete(path=path, headers=headers)
        return result
    return error_msg("delete")


def api_call(module):
    '''
    Uses playbook parameters to determine type of API request to be made
    '''
    central_api = CentralApi(module)
    action = module.params.get('action').lower()
    group_name = module.params.get('group_name')
    group_list = module.params.get('group_list')
    group_attributes = module.params.get('group_attributes')
    limit = module.params.get('limit')
    offset = module.params.get('offset')
    clone_from_group = module.params.get('clone_from_group')

    if action == "get_groups":
        result = get_groups(central_api, limit, offset)

    elif action == "get_group_mode":
        result = get_group_mode(central_api, group_list)

    elif action == "clone":
        result = clone(central_api, group_name, clone_from_group)

    elif action == "create":
        result = create_group(central_api, group_name, group_attributes)

    elif action == "update":
        result = update_group(central_api, group_name, group_attributes)

    elif action == "delete":
        result = delete_group(central_api, group_name)

    else:
        module.fail_json(changed=False,
                         msg="Unsupported action provided in playbook")

    return result


def main():
    '''
    Central Groups related parameters definitions and response handling for
    module
    '''
    module = AnsibleModule(
        argument_spec=dict(
            action=dict(required=True, type='str', choices=["get_groups",
                                                            "get_group_mode",
                                                            "clone", "update",
                                                            "create",
                                                            "delete"]),
            limit=dict(required=False, type='int', default=20),
            offset=dict(required=False, type='int', default=0),
            group_list=dict(required=False, type='list'),
            group_name=dict(required=False, type='str'),
            group_attributes=dict(required=False, type='dict'),
            clone_from_group=dict(required=False, type='str')
            ))

    success_codes = [200, 201]
    exit_codes = [304, 400, 404]
    changed = False
    if "get" not in module.params.get('action'):
        changed = True
    result = api_call(module)

    try:
        result['resp'] = json.loads(result['resp'])
    except (TypeError, ValueError):
        pass

    if result['code'] and result['code'] in success_codes:
        module.exit_json(changed=changed, msg=result['resp'],
                         response_code=result['code'])
    elif result['code'] and result['code'] in exit_codes:
        module.exit_json(changed=False, msg=result['resp'],
                         response_code=result['code'])
    else:
        module.fail_json(changed=False, msg=result['resp'],
                         response_code=result['code'])


if __name__ == '__main__':
    main()


inventory_plugins/central_inventory.py
'''
Aruba Central Inventory Plugin
'''

from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

DOCUMENTATION = r"""
    name: central_inventory
    plugin_type: inventory
    short_description: Returns Ansible inventory from YAML
    description: Returns/Generates an Ansible inventory for
                 Aruba Central from a YAML based inventory
                 plugin config file with Central credentials.
    options:
      plugin:
          description: Name of the plugin
          required: true
          choices: ["central_inventory"]
      host:
        description: Hostname for the inventory file which gets
                     generated through this plugin. Its a good practice
                     to name this parameter as "central", since
                     Aruba Central  will always be the singular host
                     for the aruba_central_role.
        required: true
      access_token:
        description: This is the Aruba Central Access token obtained
                     from the API token on Central API gateway.
                     It is valid for only 7200 seconds or 2 hours.
        required: true
      refresh_token:
        description: This is the Aruba Central Refresh token obtained
                     from the API token on Central API gateway. It is used
                     to renew an invalid access token, since access
                     token is valid only for 7200 seconds or 2 hours.
        required: true
      client_id:
        description: This is the Aruba Central Client ID obtained from
                     API token list on My Apps & Token or System Apps & Tokens
                     on Central API gateway.
        required: true
      client_secret:
        description: This is the Aruba Central Client Secret obtained from
                     API token list on My Apps & Token or System Apps & Tokens
                     on Central API gateway.
        required: true
      api_gateway:
        description: This is base url for the Aruba Central API gateway.
                     For example, for US-2 cluster the API Gateway base url
                     is apigw-prod2.central.arubanetworks.com. This base url
                     is different for different clusters.
        required: true
"""
import json
import yaml
import requests
from ansible.plugins.inventory import BaseInventoryPlugin
from ansible.errors import AnsibleError, AnsibleParserError

class InventoryModule(BaseInventoryPlugin):
    '''
    Custom inventory plugin class
    '''
    NAME = "central_inventory"

    def __init__(self):
        super(InventoryModule, self).__init__()
        self.inv_data = None
        self.plugin = None
        self.inventory_file = None
        self.host = None
        self.acc_tok = None
        self.ref_tok = None
        self.client_id = None
        self.client_sec = None
        self.api_gw = None

    def verify_file(self, path):
        """Return true/false if this is a
        valid file for this plugin to consume
        """
        valid = False
        if super(InventoryModule, self).verify_file(path):
            if path.endswith((".yml",
                              ".yaml")):
                valid = True
        return valid

    def populate(self):
        """Return the hosts and groups"""
        try:
            self.inventory.add_host(host=self.host)
            self.inventory.set_variable(
                self.host,
                "ansible_host",
                self.api_gw)
            self.inventory.set_variable(
                self.host,
                "ansible_connection",
                "httpapi")
            self.inventory.set_variable(
                self.host,
                "ansible_network_os",
                "aruba_central")
            self.inventory.set_variable(
                self.host,
                "ansible_httpapi_use_ssl",
                True)
            self.inventory.set_variable(
                self.host,
                "ansible_httpapi_central_inv_plugin",
                True)

            if self.acc_tok is not None:

                if validate_token(self.api_gw, self.acc_tok):
                    self.inventory.set_variable(
                        self.host,
                        "ansible_httpapi_central_access_token",
                        self.acc_tok)

                elif (self.ref_tok is not None and
                      self.client_id is not None and
                      self.client_sec is not None):
                    new_acc_tok, new_ref_tok = token_renew(
                        self.api_gw,
                        self.ref_tok,
                        self.client_id,
                        self.client_sec)

                    if new_acc_tok != "changeme":
                        self.inv_data["access_token"] = new_acc_tok
                        self.inv_data["refresh_token"] = new_ref_tok
                        with open(self.inventory_file, "w") as inv:
                            new_inv_data = {
                                "api_gateway": self.api_gw,
                                "host": self.host,
                                "client_id": self.client_id,
                                "plugin": self.plugin,
                                "access_token": new_acc_tok,
                                "client_secret": self.client_sec,
                                "refresh_token": new_ref_tok}
                            yaml.dump(new_inv_data, inv)
                            self.inventory.set_variable(
                                self.host,
                                "ansible_httpapi_central_access_token",
                                new_acc_tok)

                else:
                    self.inventory.set_variable(
                        self.host,
                        "ansible_httpapi_central_access_token",
                        None)

        except Exception as err:
            raise AnsibleError(
                "Use a valid credentials in the inventory "
                "plugin config file. {}".format(err)) from None

    def parse(self, inventory, loader, path, cache=True):
        '''
        Parses the inventory config source to generate
        dynamic inventory
        '''
        super(InventoryModule, self).parse(inventory, loader, path)
        self._read_config_data(path)
        self.inv_data = {}
        self.inv_data.update(self._read_config_data(path))
        try:
            self.plugin = str(self.get_option("plugin"))
            self.inventory_file = str(path)
            self.host = str(self.get_option("host"))
            self.acc_tok = str(self.get_option("access_token"))
            self.ref_tok = str(self.get_option("refresh_token"))
            self.client_id = str(self.get_option("client_id"))
            self.client_sec = str(self.get_option("client_secret"))
            self.api_gw = str(self.get_option("api_gateway"))

        except Exception as err:
            raise AnsibleParserError(
                "Check inventory plugin config for "
                "vaild parameters: {}".format(err)) from None
        self.populate()


def token_renew(api_gateway, ref_tok, client_id, client_secret):
    '''
    Renews an expired Access Token and returns a valid one
    using a valid Refresh Token, Client ID and Client Secret.
    '''
    session = requests.Session()
    path = "/oauth2/token"
    url = "https://" + api_gateway + path
    params = {"client_id": client_id, "client_secret": client_secret,
              "grant_type": "refresh_token", "refresh_token": ref_tok}
    req = requests.Request(method="POST", url=url, params=params)
    prepped = session.prepare_request(req)
    settings = session.merge_environment_settings(prepped.url, {},
                                                  None, True, None)
    response = session.send(prepped, **settings)
    if response.status_code == 200:
        new_acc_tok = json.loads(response.text)["access_token"]
        new_ref_tok = json.loads(response.text)["refresh_token"]

    elif response.status_code != 200:
        new_ref_tok = "<ENTER_VALID_REFRESH_TOKEN>"
        new_acc_tok = "<ENTER_VALID_ACCESS_TOKEN>"

    return new_acc_tok, new_ref_tok


def validate_token(api_gateway, acc_tok):
    '''
    Checks the validity of the user provided Access Token
    '''
    valid = False
    session = requests.Session()
    headers = {"Authorization": "Bearer " + acc_tok}
    path = "/configuration/v2/groups?limit=1&offset=0"
    url = "https://" + api_gateway + path
    req = requests.Request(method="GET", url=url, headers=headers)
    prepped = session.prepare_request(req)
    settings = session.merge_environment_settings(prepped.url, {}, None,
                                                  True, None)
    response = session.send(prepped, **settings)
    if response.status_code == 200:
        valid = True
    return valid
